core/types/transaction.go

```diff
@@ -61,6 +65,64 @@ type txdata struct {
 
 	// This is only used when marshaling to JSON.
 	Hash *common.Hash `json:"hash" rlp:"-"`
+
+	// Whether this is an ethereum-compatible transaction (i.e. with FeeCurrency, GatewayFeeRecipient and GatewayFee omitted)
+	EthCompatible bool `json:"ethCompatible" rlp:"-"`
+}
+
+type ethcompatibletxdata struct {
+	AccountNonce uint64          `json:"nonce"    gencodec:"required"`
+	Price        *big.Int        `json:"gasPrice" gencodec:"required"`
+	GasLimit     uint64          `json:"gas"      gencodec:"required"`
+	Recipient    *common.Address `json:"to"       rlp:"nil"` // nil means contract creation
+	Amount       *big.Int        `json:"value"    gencodec:"required"`
+	Payload      []byte          `json:"input"    gencodec:"required"`
+
+	// Signature values
+	V *big.Int `json:"v" gencodec:"required"`
+	R *big.Int `json:"r" gencodec:"required"`
+	S *big.Int `json:"s" gencodec:"required"`
+
+	// This is only used when marshaling to JSON.
+	Hash *common.Hash `json:"hash" rlp:"-"`
+
+	// Whether this is an ethereum-compatible transaction (i.e. with FeeCurrency, GatewayFeeRecipient and GatewayFee omitted)
+	EthCompatible bool `rlp:"-"`
+}
+
+func toEthCompatibleTxData(data txdata) ethcompatibletxdata {
+	return ethcompatibletxdata{
```

Comment:
i'd use named fields to create the struct. too many fields to count index and then validate it's the correct one

Reply:
Done.  I also renamed the `ethcompatibletxdata` type to make it clearer what it's for, and removed the `Hash` and `EthCompatible` fields and tags which weren't necessary for it.